server/core/region_test.go

```diff
@@ -131,6 +132,318 @@ func (*testRegionKey) TestRegionKey(c *C) {
+	c.Assert(0, Equals, bytes.Compare(overlaps[1].GetStartKey(), []byte(fmt.Sprintf("%20d", 10))))
+	c.Assert(0, Equals, bytes.Compare(item.region.GetStartKey(), []byte(fmt.Sprintf("%20d", 0))))
+
+	region = NewRegionInfo(&metapb.Region{
+		Id:       uint64(1),
+		Peers:    []*metapb.Peer{peer},
+		StartKey: []byte(fmt.Sprintf("%20d", 15)),
+		EndKey:   []byte(fmt.Sprintf("%20d", 17)),
+	}, peer)
+	overlaps, item = regions.GetOverlaps(region)
+	c.Assert(len(overlaps), Equals, 1)
+	c.Assert(0, Equals, bytes.Compare(overlaps[0].GetStartKey(), []byte(fmt.Sprintf("%20d", 10))))
+	c.Assert(0, Equals, bytes.Compare(item.region.GetStartKey(), []byte(fmt.Sprintf("%20d", 10))))
+
+	region = NewRegionInfo(&metapb.Region{
+		Id:       uint64(1),
+		Peers:    []*metapb.Peer{peer},
+		StartKey: []byte(fmt.Sprintf("%20d", 35)),
+		EndKey:   []byte(fmt.Sprintf("%20d", 45)),
+	}, peer)
+	overlaps, item = regions.GetOverlaps(region)
+	c.Assert(len(overlaps), Equals, 0)
+	c.Assert(item, IsNil)
+
+	region = NewRegionInfo(&metapb.Region{
+		Id:       uint64(1),
+		Peers:    []*metapb.Peer{peer},
+		StartKey: []byte(fmt.Sprintf("%20d", 29)),
+		EndKey:   []byte(fmt.Sprintf("%20d", 51)),
+	}, peer)
+	overlaps, item = regions.GetOverlaps(region)
+	c.Assert(len(overlaps), Equals, 2)
+	c.Assert(0, Equals, bytes.Compare(overlaps[0].GetStartKey(), []byte(fmt.Sprintf("%20d", 20))))
+	c.Assert(0, Equals, bytes.Compare(overlaps[1].GetStartKey(), []byte(fmt.Sprintf("%20d", 50))))
+	c.Assert(0, Equals, bytes.Compare(item.region.GetStartKey(), []byte(fmt.Sprintf("%20d", 20))))
+}
+
+func (*testRegionKey) TestReplaceOrAddRegion(c *C) {
+	regions := NewRegionsInfo()
+	for i := 0; i < 100; i++ {
+		peer1 := &metapb.Peer{StoreId: uint64(i%5 + 1), Id: uint64(i*5 + 1)}
+		peer2 := &metapb.Peer{StoreId: uint64((i+1)%5 + 1), Id: uint64(i*5 + 2)}
+		peer3 := &metapb.Peer{StoreId: uint64((i+2)%5 + 1), Id: uint64(i*5 + 3)}
+		region := NewRegionInfo(&metapb.Region{
+			Id:       uint64(i + 1),
+			Peers:    []*metapb.Peer{peer1, peer2, peer3},
+			StartKey: []byte(fmt.Sprintf("%20d", i*10)),
+			EndKey:   []byte(fmt.Sprintf("%20d", (i+1)*10)),
+		}, peer1)
+		regions.ReplaceOrAddRegion(region, nil, nil)
+	}
+
+	peer1 := &metapb.Peer{StoreId: uint64(4), Id: uint64(101)}
+	peer2 := &metapb.Peer{StoreId: uint64(5), Id: uint64(102)}
+	peer3 := &metapb.Peer{StoreId: uint64(1), Id: uint64(103)}
+	region := NewRegionInfo(&metapb.Region{
+		Id:       uint64(21),
+		Peers:    []*metapb.Peer{peer1, peer2, peer3},
+		StartKey: []byte(fmt.Sprintf("%20d", 184)),
+		EndKey:   []byte(fmt.Sprintf("%20d", 211)),
+	}, peer1)
+	region.learners = append(region.learners, region.voters[1])
+	region.pendingPeers = append(region.pendingPeers, region.voters[2])
+	overlaps, item := regions.GetOverlaps(region)
+	regions.ReplaceOrAddRegion(region, overlaps, item)
+	c.Assert(len(overlaps), Equals, 4)
+	c.Assert(item.region.GetID(), Equals, uint64(19))
+	c.Assert(regions.leaders[1].length(), Equals, 19)
```

Comment:
How about abstract these lines into a function to avoid repetition?

Reply:
great idea.